documentation/incremental_settings.md

```diff
@@ -0,0 +1,76 @@
+# Incremental Settings for Ansible Values
+
+The general approach in Ansible is that you provide all of the values for the
+current state of the managed object.  But this doesn't work when you have to
+manage the state of hundreds of values.  For example, when you use the
+`sysctl` module, you do not provide every single `sysctl` setting describing
+the entire state of the kernel settings.  Instead, you provide a single
+setting, which *incrementally* updates the state of all of the `sysctl`
+settings, by adding, modifying, and removing single values.
+```yaml
+- name: modify a single kernel setting
+  sysctl:
+    name: some_setting
+    value: 123  # to add/replace the value
+```
+or
+```yaml
+- name: modify a single kernel setting
+  sysctl:
+    name: some_setting
+    state: absent  # to remove the value
+```
+In the System Roles project, we had a need to generalize this for lists of
+settings and sections of settings.  We based our approach on
+[Kubernetes JSON strategic merge](https://stupefied-goodall-e282f7.netlify.app/contributors/devel/strategic-merge-patch/),
+and reuse of the Ansible `state` keyword.  We use this to manage objects which
+are represented by a `list` of `dict` objects.
+
+## Removing list items with state: absent
+
+This is the case where you want to remove some of the values from the managed
+object. Use `state: absent` in the `dict` value to remove the item named by
+`name`. For example, with `kernel_settings_sysctl` parameters:
+```yaml
+kernel_settings_sysctl:
+  - name: fs.epoll.max_user_watches
+    value: 785592
+  - name: fs.file-max
+    state: absent
+  - name: kernel.threads-max
+    state: absent
+```
+This will add/modify the parameter `fs.epoll.max_user_watches` in the managed
+object (in this case the managed object is the OS `sysctl` settings), and
+remove the the parameters `fs.file-max` and `kernel.threads-max` (in this
+case, this means the values will revert back to their system default values,
+usually upon system reboot).
```

Comment:
a note specific to this `kernel_settings` example and not to the general concept: to me it does not seem entirely right to advertise this capability (of removing values) because the result is not well defined:
- while other changes are applied immediately, this one takes effect after reboot, which is not consistent (of course, changes of bootloader parameters also require reboot, but there it is obvious, while touching a parameter that can be configured in runtime is expected to be visible immediately)
- `absent` does not really cause the underlying object (the entry in the sysctl MIB tree) to be absent. It rather sets it to some (unknown) default. This is inconsistent with the usual meaning of `absent` and also it is not clear what the result is.

The second point depends of what you actually consider the role to be managing. If you consider it a role to manage tuned profiles, then your approach makes complete sense: indeed, `state: absent` will make sure that the setting has disappeared from the tuned profile, as expected. But if it is a role that manages kernel settings, and tuned is merely a provider (implementation detail), then the point is relevant, because the role can't actually remove the settings.
In other words, `state: absent` is breaking abstraction. It is a concept valid for the provider but not so much for the general role.

Reply:
I would keep `state: absent` in the kernel_settings role, but choose a different example here to illustrate the general concept.

Reply:
addressed in https://github.com/linux-system-roles/linux-system-roles.github.io/pull/21/commits/4577d2c20cb2a14ed176486cb823442645d7088b